src/shared/codelens/goCodeLensProvider.ts

```diff
@@ -0,0 +1,179 @@
+        (await vscode.commands.executeCommand<vscode.DocumentSymbol[]>(
+            'vscode.executeDocumentSymbolProvider',
+            document.uri
+        )) ?? []
+
+    return symbols
+        .filter(symbol => isValidFuncSignature(document, symbol))
+        .map<LambdaHandlerCandidate>(symbol => {
+            return {
+                filename,
+                handlerName: `${document.uri}.${symbol.name}`,
+                manifestUri: modFile,
+                range: symbol.range,
+            }
+        })
+}
+
+/**
+ * Checks for a valid lamba function signature for Go
+ *
+ * @param document VS Code Document that contains the symbol
+ * @param symbol VS Code DocumentSymbol to analyze
+ */
+export function isValidFuncSignature(document: vscode.TextDocument, symbol: vscode.DocumentSymbol): boolean {
+    if (symbol.name === 'main') {
+        return false
+    }
+
+    // Documentation for valid lambda handler: https://docs.aws.amazon.com/lambda/latest/dg/golang-handler.html
+    // The LSP doesn't expose any low-level type-checking functionality unfortunately. If we want to perform
+    // type-checking on our lambdas, we need to write the code for that ourselves using 'executeTypeDefinitionProvider'.
+    if (symbol.kind === vscode.SymbolKind.Function && symbol.range.start.character == 0) {
+        // vscode details does not include the return type in the signature :(
+        // otherwise we would just use symbol.details instead of parsing it ourselves
+        const funcWithBody: string = document.getText(symbol.range)
+
+        const bodyStart: number = funcWithBody.search(/\{/)
+        const sigStart: number = funcWithBody.search(/\(/)
+
+        if (bodyStart === -1 || sigStart === -1) {
+            return false
+        }
+
+        const funcSig: string = stripNewLinesAndComments(funcWithBody.substring(sigStart, bodyStart))
+
+        const funcSigParts: string[] = funcSig.split(/\)/)
+        const argTypes: string[] = parseTypes(funcSigParts[0])
+        const retTypes: string[] = funcSigParts.length > 1 ? parseTypes(funcSigParts[1]) : []
+
+        if (argTypes.length > 2 || retTypes.length > 2) {
+            return false
+        }
+
+        // TODO: actually check the types to make sure they are valid
+
+        return true
+    }
+
+    return false
+}
+
+/**
+ * Finds all types of the parameter list, stripping out names if they exist.
+ * Parenthesis are stripped from the list if they exist.
+ *
+ * @param params List of parameters delimeted by commas
+ *
+ * @returns A list of types mapping 1:1 with the input
+ */
+function parseTypes(params: string): string[] {
+    params = params.replace(/\(|\)/, '')
+    const types: string[] = []
+
+    let lastType: number = 0
+    const paramParts = params.split(',')
+
+    // Names of parameters must either be all present or all absent: https://golang.org/ref/spec#Function_types
+    for (let i = 0; i < paramParts.length; i++) {
```

Comment:
These kind of loops are very are in the code base, typically we do:
```
paramParts.forEach(
```
which can also do index as the second argument